### DISCUSSION TITLE
Figure out when the render finished fully in the browser

### DISCUSSION THREAD

**Discussion by njfamirm** [lit/lit#4756] [Q&A] [answered]
Hi everyone

I'm use lit web component for render a message list in the chat, and want scroll to button of them when the render fully complete and the element height is set correctly

Using `updateComplete` is not solve my problem

```ts
protected override updated(changedProperties: PropertyValues<this>): void {
    super.updated(changedProperties);

    if (changedProperties.has('messageList')) {
      this.updateComplete.then(() => this.scrollToBottom__(false));
    }
  }

  /**
   * Scroll to the bottom.
   */
  private scrollToBottom__(): void {
    this.scrollElement__.scrollTo({
      top: this.scrollElement__.scrollHeight,
    });
  }
```

**Comment by sorvell** [comment]
Lit renders all elements using the [microtask queue](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide) so you can wait for some "task" like `requestAnimationFrame` before scrolling. I'd suggest:

```ts
async private scrollToBottom__(): void {
  await new Promise(requestAnimationFrame);
  this.scrollElement__.scrollTo({
    top: this.scrollElement__.scrollHeight,
  });
}
```

**Reply by njfamirm** [reply]
thanks

but the `await new Promise(requestAnimationFrame);` is not resolved

```ts
protected override updated(changedProperties: PropertyValues<this>): void {
    super.updated(changedProperties);

    // scroll to the bottom when the message list changes
    if (changedProperties.has('messageList')) {
      // TODO: better handle this
      this.updateComplete.then(this.scrollToBottom__.bind(this, {smooth: false}))
    }
  }

  /**
   * Scroll to the bottom.
   */
  private async scrollToBottom__(options: {smooth: boolean}): Promise<void> {
    await waitForAnimationFrame()
    this.logger_.logMethodArgs?.('scrollToBottom__', {options, scrollHeight: this.scrollElement__.scrollHeight});

    this.scrollElement__.scrollTo({
      top: this.scrollElement__.scrollHeight,
      behavior: options.smooth ? 'smooth' : 'auto',
    });
  }
```

also this code still not work

**Reply by njfamirm** [reply]
oh! solved, my problem is from external animation showing!

**Reply by njfamirm** [reply, marked as answer]
no, your solution in not enough for some situation and I'm found the correct way:

```ts
requestAnimationFrame(() => {
  setTimeout(() => {}, 0);
});
```


[Nicholas Tower](https://stackoverflow.com/users/3794812/nicholas-tower)'s Response on [StackOverflow](https://stackoverflow.com/a/47184426)

> Running code immediately after a repaint maximizes the chance that the DOM has been fully calculated, and thus minimizes the chance that querying the dom will cause a costly reflow. If you're not querying the dom anyway, then this isn't something you need to worry about.
  requestAnimationFrame will schedule code to run immediately before the repaint, which is close to where you want to be but not quite. So then doing a setTimeout of 0 (or setImmediate on browsers that support it) will execute code as soon as it can after that. This won't guarantee that your code is the first thing to run after the repaint, but it's the best you can do given the apis at your disposal.
